File: python/pythagorean_triples/pyth.py

class Triple:

    def __init__(self):
        self.a = None
        self.b = None
        self.c = None
        self.valid = True

def search(a, b, start, end, inc):

    for c in range(start, end + 1, inc):
        if (c*c == a*a + b*b):
            return c

    return 0;

def remove_multiples(found, num):

    for i in range(0, num):
        if found[i].valid:
            k = 2
            for j in range(i+1, num):
                if found[j].valid:
                    if ( found[i].a * k == found[j].a and
                         found[i].b * k == found[j].b and
                         found[i].c * k == found[j].c ):
                      found[j].valid = False
                      k += 1

def main():

    MIN = 1
    MAX = 2000
    MAX_TRIPLES = 4000
    found = [Triple() for i in range(MAX_TRIPLES)]

    # c**2 = a**2 + b**2

    found_triples = 0
    for a in range(MIN, MAX):
        for b in range(a, MAX): 
            c = search (a, b, b+1, MAX, 1)
            if (c):
                #print('{}, {}, {}, {}'.format(found_triples, a, b, c))
                found[found_triples].a = a
                found[found_triples].b = b
                found[found_triples].c = c
                found_triples += 1

    print('Found {}'.format(found_triples))
    remove_multiples(found, found_triples)

    final_triples = 0
    for i in range(0, found_triples):
        if found[i].valid:
            print('{}, {}, {}, {}'.format(i, found[i].a, found[i].b, found[i].c))
            final_triples += 1

    print('Final {}'.format(final_triples))

if __name__ == "__main__":
    main()

File: go/pythagorean_triples/pyth.go
package main

import "fmt"

type triple struct {
    a int
    b int
    c int
    valid bool
}


func search(a int, b int, start int, end int, inc int) int {

    for c := start; c <= end; c += inc {
        if c*c == a*a + b*b {
            return c
        }
    }

    return 0
}

func remove_multiples(found []triple, num int) {
    var k int
    for i := 0; i < num; i++ {
        if found[i].valid {
            k = 2
            for j := i+1; j < num; j++ {
                if found[j].valid {
                    if ( found[i].a * k == found[j].a &&
                         found[i].b * k == found[j].b &&
                         found[i].c * k == found[j].c ){
                      found[j].valid = false
                      k++
                    }
                }
            }
        }
    }
}

func main() {

    const MIN int = 1
    const MAX int = 2000
    const MAX_TRIPLES int = 4000
    var found [MAX_TRIPLES]triple

    // c**2 = a**2 + b**2

    var c int
    var found_triples = 0
    for a := MIN; a < MAX; a++ {
        for b := a; b < MAX; b++ { 
            c = search (a, b, b+1, MAX, 1)
            
            if c != 0 {
                //fmt.Printf("%d, %d, %d, %d\n", found_triples, a, b, c)
                found[found_triples].a = a
                found[found_triples].b = b
                found[found_triples].c = c
                found[found_triples].valid = true
                found_triples++
            }
        }
    }

    fmt.Printf("Found %d\n", found_triples)
    remove_multiples(found[:], found_triples)

    var final_triples = 0
    for i := 0; i < found_triples; i++ {
        if found[i].valid {
            fmt.Printf("%d, %d, %d, %d\n", i, found[i].a, found[i].b, found[i].c)
            final_triples++
        }
    }
    fmt.Printf("Final %d\n", final_triples)
}


